src/devices/sound/qsoundhle.cpp

```diff
@@ -82,74 +68,120 @@ void qsound_hle_device::device_start()
+		save_item(NAME(m_voice[j].rate),j);
+		save_item(NAME(m_voice[j].loop_len),j);
+		save_item(NAME(m_voice[j].end_addr),j);
+		save_item(NAME(m_voice[j].volume),j);
+		save_item(NAME(m_voice[j].echo),j);
+	}
+
+	for(int j=0;j<3;j++) // ADPCM voices
+	{
+		save_item(NAME(m_adpcm[j].start_addr),j);
+		save_item(NAME(m_adpcm[j].end_addr),j);
+		save_item(NAME(m_adpcm[j].bank),j);
+		save_item(NAME(m_adpcm[j].volume),j);
+		save_item(NAME(m_adpcm[j].flag),j);
+		save_item(NAME(m_adpcm[j].cur_vol),j);
+		save_item(NAME(m_adpcm[j].step_size),j);
+		save_item(NAME(m_adpcm[j].cur_addr),j);
+	}
 
-	for (int adr = 0x80; adr < 0x90; adr++)
-		write_data(adr, 0x120);
+	for(int j=0;j<19;j++) // PCM voices
+	{
+		save_item(NAME(m_voice_pan[j]),j);
+	}
 
-	// state save
-	for (int i = 0; i < 16; i++)
+	// QSound registers
+	save_item(NAME(m_echo.end_pos));
+	save_item(NAME(m_echo.feedback));
+	save_item(NAME(m_echo.length));
+	save_item(NAME(m_echo.last_sample));
+	save_item(NAME(m_echo.delay_line));
+	save_item(NAME(m_echo.delay_pos));
+
+	for(int j=0;j<2;j++)  // left, right
 	{
-		save_item(NAME(m_channel[i].reg), i);
-		save_item(NAME(m_channel[i].lvol), i);
-		save_item(NAME(m_channel[i].rvol), i);
+		save_item(NAME(m_filter[j].tap_count),j);
+		save_item(NAME(m_filter[j].delay_pos),j);
+		save_item(NAME(m_filter[j].table_pos),j);
+		save_item(NAME(m_filter[j].taps),j);
+		save_item(NAME(m_filter[j].delay_line),j);
+
+		save_item(NAME(m_alt_filter[j].tap_count),j);
+		save_item(NAME(m_alt_filter[j].delay_pos),j);
+		save_item(NAME(m_alt_filter[j].table_pos),j);
+		save_item(NAME(m_alt_filter[j].taps),j);
+		save_item(NAME(m_alt_filter[j].delay_line),j);
+
+		save_item(NAME(m_wet[j].delay),j);
+		save_item(NAME(m_wet[j].volume),j);
+		save_item(NAME(m_wet[j].write_pos),j);
+		save_item(NAME(m_wet[j].read_pos),j);
+		save_item(NAME(m_wet[j].delay_line),j);
+
+		save_item(NAME(m_dry[j].delay),j);
+		save_item(NAME(m_dry[j].volume),j);
+		save_item(NAME(m_dry[j].write_pos),j);
+		save_item(NAME(m_dry[j].read_pos),j);
+		save_item(NAME(m_dry[j].delay_line),j);
 	}
+
+	save_item(NAME(m_state));
+	save_item(NAME(m_next_state));
+	save_item(NAME(m_delay_update));
+	save_item(NAME(m_state_counter));
+	save_item(NAME(m_ready_flag));
+	save_item(NAME(m_data_latch));
+	save_item(NAME(m_out));
 }
 
+//-------------------------------------------------
+//  rom_region - return a pointer to the device's
+//  internal ROM region
+//-------------------------------------------------
+
+const tiny_rom_entry *qsound_hle_device::device_rom_region() const
+{
+	return ROM_NAME( qsound_hle );
+}
 
 //-------------------------------------------------
 //  device_reset - device-specific reset
 //-------------------------------------------------
 
 void qsound_hle_device::device_reset()
 {
-	for (qsound_channel &ch : m_channel)
-		std::fill(std::begin(ch.reg), std::end(ch.reg), 0U);
+	m_ready_flag = 0;
+	m_out[0] = m_out[1] = 0;
+	m_state = 0;
+	m_state_counter = 0;
```

Comment:
Resetting the device would cause the DSP to jump to address zero, causing it to zero its registers and reinitialise.